### ISSUE TITLE
Create linux binary of Oak Functions

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#4182]
This will be a stand-alone binary (including a gRPC server) that will eventually be packaged as a Docker image to run on Oak Containers.

cc @pmcgrath17

**Comment by tiziano88** [comment]
In the linux version of Oak Functions we would like to leverage async execution and threads. OTOH we don't have either of these in the Oak Restricted Kernel.

I'll look around for an async executor that is no_std compatible (tokio is not).

If anyone knows of one, please let me know.

**Comment by tiziano88** [comment]
As next step, I am planning to create a pure Rust object with the methods corresponding to the [existing micro RPC interface](https://github.com/project-oak/oak/blob/4d401f81e014aa34b5048a1f39ff7f15e79ba03a/oak_functions_service/proto/oak_functions.proto), as a separate crate. Then we can wrap that object in the micro RPC interface for Oak Restricted Kernel, and in a gRPC server for Oak Containers.

**Comment by jul-sh** [comment]
> I'll look around for an async executor that is no_std compatible (tokio is not).
> 
> If anyone knows of one, please let me know.

Why do we require no_std if we're targeting linux?

**Comment by jul-sh** [comment]
There's some overlap between what this issue proposes, and the previous oak functions linux implementation (currently being removed in #3738). 

The differences appears to be that
a) previous implementation is two binaries that communicate over a socket (likely less performant)
b) we want to leverage multithreading presumably to execute multiple workloads in parralel. 

For a), it would be faster to re-use the existing implementation for now. Have we considered this option? I'm not convinced that the performance bottleneck introduced by the single channel is massive enough to justify fixing it right away. We can always improve performance later. :)

**Comment by tiziano88** [comment]
Thanks @jul-sh , you are correct that there is some overlap.

I did initially consider reusing the existing split binary version, but I decided against it, since in that case the enclave binary would still be single threaded (your point b).

Additionally, we have already refactored the core logic of Oak Functions as a library (and wrapped in micro RPC) (#4199 continues that work). So having a type safe Rust library that can just be used as part of a larger binary seems a better solution than having to use a socket to communicate; I'm mostly not concerned about performance (though it's also a factor, especially for large chunks of data that would have to be copied multiple times), rather about type safety and maintainability.
